dm/master/openapi.go

```diff
@@ -0,0 +1,153 @@
+// Copyright 2021 PingCAP, Inc.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+// this file implement all of the APIs of the DataMigration service.
+
+package master
+
+import (
+	"context"
+	"fmt"
+	"net/http"
+	"os"
+
+	"github.com/deepmap/oapi-codegen/pkg/middleware"
+	"github.com/getkin/kin-openapi/openapi3"
+	"github.com/labstack/echo/v4"
+	echomiddleware "github.com/labstack/echo/v4/middleware"
+	"go.uber.org/zap"
+
+	"github.com/pingcap/dm/openapi"
+	"github.com/pingcap/dm/pkg/log"
+	"github.com/pingcap/dm/pkg/terror"
+)
+
+const (
+	docBasePath     = "/api/v1/docs"
+	docJSONBasePath = "/api/v1/dm.json"
+)
+
+// InitOpenAPIHandles init openapi handlers.
+func (s *Server) InitOpenAPIHandles() {
+	swagger, err := openapi.GetSwagger()
+	checkServerErr(err)
+	swagger.AddServer(&openapi3.Server{URL: fmt.Sprintf("http://%s", s.cfg.AdvertiseAddr)})
+	swaggerJSON, err := swagger.MarshalJSON()
+	checkServerErr(err)
+	docMW, err := openapi.NewSwaggerDocUI(openapi.NewSwaggerConfig(docBasePath, docJSONBasePath, ""), swaggerJSON)
+	checkServerErr(err)
+	e := echo.New()
+	// inject err handler
+	e.HTTPErrorHandler = terrorHTTPErrorHandler
+	// middlewares
+	e.Pre(docMW)
+	// set logger
+	logger := log.L().Logger
+	logger = logger.With(zap.String("component", "openapi"))
+	e.Use(openapi.ZapLogger(logger))
+	e.Use(echomiddleware.Recover())
+	// disables swagger server name validation. it seems to work poorly
+	swagger.Servers = nil
+	// use our validation middleware to check all requests against the OpenAPI schema.
+	e.Use(middleware.OapiRequestValidator(swagger))
+	openapi.RegisterHandlers(e, s)
+	s.echo = e
+}
+
+// redirectRequestToLeader is used to redirect the request to leader.
+// because the leader has some data in memory, only the leader can process the request.
+func (s *Server) redirectRequestToLeader(ctx context.Context) (needRedirect bool, host string, err error) {
```

Comment:
Will this function be used in next PR?

Reply:
yes, it will be used like [this](https://github.com/Ehco1996/dm/blob/7c01c9cacf60d882548e0bbcc2486d444c9bace2/dm/master/openapi.go#L109)